federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -516,7 +519,17 @@ protected UserModel importUserFromLDAP(KeycloakSession session, RealmModel realm
 
         UserModel imported = null;
         if (model.isImportEnabled()) {
-            imported = session.userLocalStorage().addUser(realm, ldapUsername);
+            // Search if there is already an existing user, which means the username might have changed in LDAP without Keycloak knowing about it
+            UserModel existingLocalUser = session.userLocalStorage()
+                    .searchForUserByUserAttributeStream(realm, LDAPConstants.LDAP_ID, ldapUser.getUuid()).findFirst().orElse(null);
+            if(existingLocalUser != null){
+                imported = session.userLocalStorage().getUserByUsername(realm, existingLocalUser.getUsername());
```

Comment:
Won't it be easier to use just `imported = existingLocalUser;` on this line? I think there is no need to trigger again `session.userLocalStorage().getUserByUsername` as that should always return same user as `existingLocalUser` if I am not mistaken? So this would just mean another SQL query to the DB.

Reply:
Good point, have overseen this and corrected it